File: arduino/PackCommunication.cpp
#include "PackCommunication.h"

PackCommunication Paco;

const char PackCommunication::SFLAG = 0x30;
const char PackCommunication::EFLAG = 0x31;
const char PackCommunication::ESCC  = 0x32;

PackCommunication::PackCommunication() :
	ss      (nullptr),
	byteT   (1000000/9600),
	flushing(false) {}
PackCommunication::~PackCommunication(){}

bool PackCommunication::begin(HardwareSerial &serial, bool _flushing)
{
	ss       = &serial;
	byteT    = 1000000/9600;
	flushing = _flushing;
	return ss;
}

void PackCommunication::send(char *bytes, int size_b)
{
	if (size_b < 1)	return;

	ss->write(SFLAG);
	//ss->flush();
	char c;
	for (int i=0;i<size_b;i++) {
		c = (bytes[i]);
		if (c == SFLAG || c == EFLAG || c == ESCC) {
			ss->write(ESCC);
			//ss->flush();
		}
		ss->write(c);
		//ss->flush();
	}
	ss->write(EFLAG);
	if (flushing)
		ss->flush();
}

bool PackCommunication::recv(char *bytes, int &size_b)
{
	size_b = 0;
	if (ss->available() < 1) return false;

	char c;
	if (ss->readBytes(&c, 1) < 1 || c != SFLAG)	return false;
	do {
		while (ss->available() < 1) delayMicroseconds(byteT);
		ss->readBytes(&c, 1);
		if (c == EFLAG) break;
		if (c == ESCC) {
			while (ss->available() < 1) delayMicroseconds(byteT);
			ss->readBytes(&c, 1);
		}
		bytes[size_b++] = c;
	} while (1);

	return true;
}

bool PackCommunication::recv_wait(char *bytes, int &size_b)
{
	while (!recv(bytes, size_b)) {
		delayMicroseconds(byteT);
	} 
	return true;
}

File: python/PackCommunication.py
#!/usr/bin/env python

import serial
import time
import msvcrt
import sys

class PackCommunication:

	def __init__(self, path, baud_rate):
		self.SFLAG = b'0'
		self.EFLAG = b'1'
		self.EESC  = b'2'
		self.byteT = 1.0/baud_rate

		self.s     = serial.Serial()
		self.error = ""
		try:
			self.s.port     = path
			self.s.baudrate = baud_rate
			self.s.timeout  = 0

			self.s.open()
			time.sleep(2.000)			# arduino reset
		except serial.serialutil.SerialException:
			self.error = path + " not found"

	def opened(self):
		return self.s.is_open

	def send(self, string):
		self.s.write(self.SFLAG)
		
		for c in string:
			if c.encode() == self.EFLAG or c.encode() == self.SFLAG or c.encode() == self.EESC:
				self.s.write(self.EESC)
			self.s.write(c.encode())
		
		self.s.write(self.EFLAG)
		self.s.flush()

	def send_bytes(self, _bytes):
		self.s.write(self.SFLAG)
		
		for c in _bytes:
			c = bytes([c])
			if c == self.EFLAG or c == self.SFLAG or c == self.EESC:
				self.s.write(self.EESC)
			self.s.write(c)

		self.s.write(self.EFLAG)
		self.s.flush()

	def recv(self):
		buff = bytearray()
		if (self.s.in_waiting < 2):
			return buff
		
		byte = self.s.read()
		if (byte != self.SFLAG):
			return buff

		while True:
			byte = self.s.read()
			if len(byte) == 0:
				continue
			elif byte == self.EFLAG:
				break
			elif byte == self.EESC:
				byte = self.s.read()
				while len(byte) < 1:
					byte = self.s.read()
			
			buff.extend(byte)

		return buff

	def recv_wait(self):
		while (self.s.in_waiting < 2):
			time.sleep(self.byteT)
		return self.recv()

# Program

if __name__ == "__main__":
	print("Example 1")

	paco = PackCommunication("COM3", 115200)
	if not paco.opened():
		print("Error : ", paco.error)
		exit(1)

	while True:
		paco.send("Hello World")
		print(paco.recv_wait().decode("utf-8") )

	